### ISSUE TITLE
bug: `ion-reorder` complete() method moves injected Angular comments used by *ngFor 

### ISSUE THREAD

**Issue by byronigoe** [ionic-team/ionic-framework#25060]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [X] v5.x
- [X] v6.x

### Current Behavior

After the first item of a list is reordered, additional items are inserted at locations other than the bottom.

The root cause is that when doReorder calls `ev.detail.complete();` Ionic's DOM manipulation causes the Angular binding comment to move relative to the list items. New DOM elements are inserted by Angular at the location of this comment.
```
<!--bindings={
  "ng-reflect-ng-for-of": "Item 0,Item 1,Item 2,Item 3"
}-->
```

### Expected Behavior

New items pushed onto the list array should be added to the bottom.

Specifically, the location of Angular's comment should remain at the end of the items generated by ngFor after Ionic reorders the DOM elements.

### Steps to Reproduce

1. Use `@ionic/angular`.
1. Create an `ion-reorder-group` with a list of items generated by `ngFor`.
1. Drag the first item down (e.g. to the bottom of the list).
1. Now add a new item to the array referenced by `ngFor`.

### Code Reproduction URL

https://stackblitz.com/edit/ionic6-angular13-chauqs

### Ionic Info

The StackBlitz is using Angular 13.3.1 and @ionic/angular 6.0.14, but the same behavior is present on 5.5

### Additional Information

The workaround I'm using for now is to reject the reorder operation by calling `ev.detail.complete(false);`, so the DOM does not change, rely on Angular to synchronize the DOM with my array, and not have any animation on the bounce back.

**Comment by sean-perkins** [comment]
@byronigoe after discussion with the team, this issue is not a bug, but an implementation detail. 

If you update the implementation of `doReorder` to:
```ts
doReorder(ev: any) {
  console.log('Dragged from index', ev.detail.from, 'to', ev.detail.to);
  this.items = ev.detail.complete(this.items);
  this.dump();
}
```

You will notice the rendering is correct, the items are pushed to the end of the collection and your `json` value is in sync with the reorder events. 

Updated stackblitz: https://stackblitz.com/edit/ionic6-angular13-ypwrbv

The `complete` method expects the list of items, so that elements are reordered in the correct location. 

I am going to close out this issue. Thanks!

**Comment by byronigoe** [comment]
Passing the list of items to complete() is fine when you want Ionic to change your list, but complete() can also take "true" as a parameter and it moves the DOM element only.

I need to do the actual reordering myself. The example in the StackBlitz uses a simple list, whereas my actual use case is reordering sibling branches at one level of a larger tree.

Please reopen this issue.

**Comment by sean-perkins** [comment]
@byronigoe I've reopened the issue, but will need additional detail as to why you need to perform the reordering logic internal to your application. 

From a brief discovery, `insertBefore` isn't aware of non-HTML nodes, so controlling the insert position is likely more complex. This behavior has existed since pre-Ionic 4 days, so I am optimistic that understanding why your application needs to own the reordering, will help identify the solution.

**Comment by byronigoe** [comment]
My data is a tree, represented as an array of objects with each element referencing its parent. For example:
```json
[
  { id: 0, name: "Grandparent A" },
  { id: 1, name: "Parent A", parent: "Grandparent A" },
  { id: 2, name: "Child A", parent: "Parent A" },
  { id: 3, name: "Child B", parent: "Parent A" },
  { id: 4, name: "Child C", parent: "Parent A" },
  { id: 5, name: "Parent B", parent: "Grandparent A" },
  { id: 6, name: "Child D", parent: "Parent B" },
  { id: 7, name: "Child E", parent: "Parent B" }
]
```

In my app, users can select a node (e.g. Grandparent A), and reorder the child-nodes (e.g. move Parent A below Parent B). The descendants of the child-nodes need to move too. The result of this example operation would be:
```json
[
  { id: 0, name: "Grandparent A" },
  { id: 5, name: "Parent B", parent: "Grandparent A" },
  { id: 6, name: "Child D", parent: "Parent B" },
  { id: 7, name: "Child E", parent: "Parent B" },
  { id: 1, name: "Parent A", parent: "Grandparent A" },
  { id: 2, name: "Child A", parent: "Parent A" },
  { id: 3, name: "Child B", parent: "Parent A" },
  { id: 4, name: "Child C", parent: "Parent A" }
]
```

**Comment by liamdebeasi** [comment]
@byronigoe Can you provide a GitHub repo of your actual use case?

**Comment by liamdebeasi** [comment]
Can you also clarify why `ev.detail.complete(false)` does not work for your use case? That should disable Ionic's internal reorder of the DOM nodes. From there, your application reorders the array which causes Angular to correctly re-render the list.

**Comment by byronigoe** [comment]
> @byronigoe Can you provide a GitHub repo of your actual use case?

No, and this is not necessary. I've clearly explained and demonstrated the problem.


> Can you also clarify why `ev.detail.complete(false)` does not work for your use case?

This workaround does work for my use case, as I mentioned in the original submission. I have submitted this bug so it can be fixed for others.

**Comment by liamdebeasi** [comment]
I took a closer look at our code, and the reorder component is working as intended. There are a few different options when calling `complete()`:

1. Pass in no parameters. Ionic will automatically re-order the DOM nodes. For simple use cases this should be sufficient. If you have the rendering of the DOM nodes linked up to the state of an array, this can cause issues. In the case of your sample app, Angular was trying to re-render the list items at the same time that Ionic was trying to reorder them.
2. Pass in the array that renders the list. Ionic will automatically re-order the DOM nodes and sort and return the array. This is preferred when you have the rendering of a list hooked up to the state of an array.
3. Pass in `false`. Ionic will neither re-order the DOM nor sort and return the array. This is essentially ejecting from Ionic's default behavior when there needs to be additional custom logic here. This is the recommended solution for your use case as it causes all of the rendering to go through Angular's rendering system.

I updated the documentation in https://github.com/ionic-team/ionic-framework/pull/25086 to make this more clear.

I am going to close this issue. Thanks!
